spongefish/src/pattern/interaction_pattern.rs

```diff
@@ -0,0 +1,190 @@
+use core::fmt::Display;
+
+use thiserror::Error;
+
+use super::{interaction::Hierarchy, Interaction, Kind};
+
+/// Abstract transcript containing prover-verifier interactions
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug, Default)]
+pub struct InteractionPattern {
+    interactions: Vec<Interaction>,
+}
+
+/// Errors when validating a transcript.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash, Error)]
+pub enum TranscriptError {
+    #[error("Missing Begin for {end} at {position}")]
+    MissingBegin { position: usize, end: Interaction },
+    #[error(
+        "Invalid kind {interaction} at {interaction_position} for {begin} at {begin_position}"
+    )]
+    InvalidKind {
+        begin_position: usize,
+        begin: Interaction,
+        interaction_position: usize,
+        interaction: Interaction,
+    },
+    #[error("Mismatch {begin} at {begin_position} for {end} at {end_position}")]
+    MismatchedBeginEnd {
+        begin_position: usize,
+        begin: Interaction,
+        end_position: usize,
+        end: Interaction,
+    },
+    #[error("Missing End for {begin} at {position}")]
+    MissingEnd { position: usize, begin: Interaction },
+}
+
+impl InteractionPattern {
+    pub fn new(interactions: Vec<Interaction>) -> Result<Self, TranscriptError> {
+        let result = Self { interactions };
+        result.validate()?;
+        Ok(result)
+    }
+
+    #[must_use]
+    #[allow(clippy::missing_const_for_fn)] // False positive
+    pub fn interactions(&self) -> &[Interaction] {
+        &self.interactions
+    }
+
+    /// Generate a unique identifier for the protocol.
+    ///
+    /// It is created by taking the SHA3 hash of a stable unambiguous
+    /// string representation of the transcript interactions.
+    // TODO: A more neutral implementation would use ASN.1 DER.
+    #[must_use]
+    pub fn domain_separator(&self) -> [u8; 32] {
```

Comment:
Following from yesterday, would maybe change the name to something like `protocol_interaction_pattern_identifier` to reflect that this is something that can be used as part of the domain_separator but it's not the whole domain separator.

Reply:
Agree. I'd like a shorter name though, something of similar complexity as `domain_separator`. Maybe `pattern_hash` ?

Reply:
Good for me!